### PR TITLE
Tail lock revisited

### PR DESCRIPTION
**Pull request by ceyonur**
## Why this should be merged

There is a potential race between txindexer and state sync (ResetToStateSyncedBlock/repairTxIndexTail), where reading can occur before repairTxIndexTail (which can increment the tail) and txindexer can work on an outdate tail (less than actual). 

## How this works

Uses txindexer's `run` function after grabbing lock and then do the actual db read for tail for `run`. 

## How this was tested

Not sure how to test this race

## How is this documented

no need

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -2183,18 +2183,12 @@ func (bc *BlockChain) CacheConfig() *CacheConfig {
...
 
-	tailP := rawdb.ReadTxIndexTail(bc.db)
-	var tailV uint64
-	if tailP != nil {
-		tailV = *tailP
-	}
-
-	if newTail > tailV {
-		log.Info("Repairing tx index tail", "old", tailV, "new", newTail)
+	if currentTail := rawdb.ReadTxIndexTail(bc.db); currentTail == nil || newTail > *currentTail {
```
Comment (ARR4N) [inline core/blockchain.go:2190]: Question: as far as I can tell this is just a refactor with the same logic; is that your intention?

Comment: I think the variable `currentTail` could just be `curr` (or even `tail`) to make this more succinct and easier to read. The fact that it's a tail is implied by the RHS of the assignment, and it's only used on two lines afterwards so verbosity is unnecessary:

```go
if curr := rawdb.ReadTxIndexTail(bc.db); curr == nil || *curr < newTail {
// or
if tail := rawdb.ReadTxIndexTail(bc.db); tail == nil || *tail < newTail {
```

Afterthought: my using `curr < new` instead of `new > curr` just kinda happened as it felt more natural to frame everything in terms of `curr`. Entirely subjective, and up to you to decide if you think I'm nuts :)

File: core/txindexer.go [hunk 2]
```diff
@@ -99,33 +96,27 @@ func (indexer *txIndexer) run(tail *uint64, head uint64, stop chan struct{}, don
 	}
 
 	// Defensively ensure tail is not nil.
-	if tail == nil {
-		tail = new(uint64)
+	tailValue := uint64(0)
```
Comment (ARR4N) [inline core/txindexer.go:99]: Why the change here? It's effectively the same thing.

File: core/txindexer.go [hunk 3]
```diff
@@ -145,7 +136,12 @@ func (indexer *txIndexer) loop(chain *BlockChain) {
 		done = make(chan struct{})
 		lastHead = head.Number.Uint64()
 		indexer.chain.wg.Add(1)
-		go indexer.run(rawdb.ReadTxIndexTail(indexer.db), head.Number.Uint64(), stop, done)
+		go func() {
```
Comment (ARR4N) [inline core/txindexer.go:139]: Am I correct in understanding that this (and the similar go routine below) is where the fix occurs, because the call to `rawdb.ReadTxIndexTail` is now inside the lock?

Does `ResetToStateSyncedBlock` honour the same lock?

File: core/txindexer.go (hunk 2 above)
Comment (ceyonur) [reply core/txindexer.go:99]: it avoids overriding pointer

File: core/txindexer.go (hunk 3 above)
Comment (ceyonur) [reply core/txindexer.go:139]: yes, `ResetToStateSyncedBlock` uses `repairTxIndexTail` which holds the same lock

File: core/blockchain.go (hunk 1 above)
Comment (ceyonur) [reply core/blockchain.go]: yes it's a refactor

I applied your suggestion, thanks.

File: core/txindexer.go [hunk 4]
```diff
@@ -159,15 +155,19 @@ func (indexer *txIndexer) loop(chain *BlockChain) {
 				stop = make(chan struct{})
 				done = make(chan struct{})
 				indexer.chain.wg.Add(1)
-				go indexer.run(rawdb.ReadTxIndexTail(indexer.db), head.Block.NumberU64(), stop, done)
+				go func() {
```
Comment (ARR4N) [inline core/txindexer.go:158]: If this must always be the same as the equivalent go routine then consider abstracting it into shared code, otherwise all LGTM.

**Review by ARR4N — APPROVED** [review]
